### DISCUSSION TITLE
Simple code does not run

### DISCUSSION THREAD

**Discussion by rahav** [contiki-ng/contiki-ng#2734] [Q&A] [answered]
After five days of debugging #2726, the very small code below simply refuses to run on nRF52840 Dongle. I am desperately out of ideas and will immensely appreciate any direction or ideas. At this time I am probably doing something really naive, that I am missing.

The code will run if I remove Contiki Shell from `Makefile`. My confidence got shaken, when it started to work with Shell included, given any of the following:
- I comment any single members of `device_configuration`
- Print something (even "\n\n" but not "\n") before or after the long `printf`
- Split the long printf to two, and move at least 4 variables to be printed by the first.

````
#include "contiki.h"
#include <stdio.h>	/* printf() */

struct device_configuration {
	uint8_t type;
	char type_desc[10];
	int16_t state_range_start;
	int16_t state_range_end;
	int16_t ini_value;
	uint8_t send_threshold_events;
	uint8_t change_state_on_threshold_event;
	int16_t first_threshold_event_in;
} dc;

const struct device_configuration devices_db[] = { 
	{86, "Water", 0,  1,   0, 0, 0, 207},
	{84, "Timer", 0,  9,   0, 0, 0, 207},
	{83, "Soil", 10, 90, 207, 1, 1,   5}
 };

/*---------------------------------------------------------------------------*/
PROCESS(boot_process, "Boot process");
AUTOSTART_PROCESSES(&boot_process);

/*---------------------------------------------------------------------------*/
PROCESS_THREAD(boot_process, ev, data)
{
	static struct etimer timer;

	PROCESS_BEGIN();
	etimer_set(&timer, CLOCK_SECOND * 1);

	while(1) {
		for (int i = 0; i < 3; ++i) {
			// printf("Hi\n");   // adding this printf, causes it to work
			// printf("\n");     // printing one \n does not cause it to work, but printing two does

			printf("type = %d, type_desc = %s, state_range_start = %i, state_range_end = %i, ini_value = %i, send_threshold_events = %d, change_state_on_threshold_event = %d, first_threshold_event_in = %i\n",
				devices_db[i].type,
				devices_db[i].type_desc,
				devices_db[i].state_range_start,
				devices_db[i].state_range_end,
				devices_db[i].ini_value,
				devices_db[i].send_threshold_events,
				devices_db[i].change_state_on_threshold_event,
				devices_db[i].first_threshold_event_in
				);
		}
		printf("\n");

		PROCESS_WAIT_EVENT_UNTIL(etimer_expired(&timer));
		etimer_reset(&timer);
	}

	PROCESS_END();
}

````

**Comment by pjonsson** [comment]
This is latest develop and the corresponding `contiker/contiki-ng:latest`, but I don't think it makes a significant difference for your example.

I removed the code in `examples/hello-world/hello-world.c` and pasted the code into that file. After that, I ran:
```
$ make TARGET=nrf52840 BOARD=dongle hello-world.dfu-upload PORT=/dev/ttyACM0
<many lines>

|===============================================================|
|##      ##    ###    ########  ##    ## #### ##    ##  ######  |
|##  ##  ##   ## ##   ##     ## ###   ##  ##  ###   ## ##    ## |
|##  ##  ##  ##   ##  ##     ## ####  ##  ##  ####  ## ##       |
|##  ##  ## ##     ## ########  ## ## ##  ##  ## ## ## ##   ####|
|##  ##  ## ######### ##   ##   ##  ####  ##  ##  #### ##    ## |
|##  ##  ## ##     ## ##    ##  ##   ###  ##  ##   ### ##    ## |
| ###  ###  ##     ## ##     ## ##    ## #### ##    ##  ######  |
|===============================================================|
|You are not providing a signature key, which means the DFU     |
|files will not be signed, and are vulnerable to tampering.     |
|This is only compatible with a signature-less bootloader and is|
|not suitable for production environments.                      |
|===============================================================|

Zip created at build/nrf52840/dongle/nrf52840_dfu_image.zip
nrfutil dfu usb-serial -p /dev/ttyACM0 -pkg build/nrf52840/dongle/nrf52840_dfu_image.zip
  [####################################]  100%          
Device programmed.
$ make TARGET=nrf login PORT=/dev/ttyACM0
rlwrap ../../tools/serial-io/serialdump -b115200 /dev/ttyACM0
connecting to /dev/ttyACM0 [OK]
type = 86, type_desc = Water, state_range_start = 0, state_range_end = 1, ini_value = 0, send_threshold_events = 0, change_state_on_threshold_event = 0, first_threshold_event_in = 207
type = 84, type_desc = Timer, state_range_start = 0, state_range_end = 9, ini_value = 0, send_threshold_events = 0, change_state_on_threshold_event = 0, first_threshold_event_in = 207
type = 83, type_desc = Soil, state_range_start = 10, state_range_end = 90, ini_value = 207, send_threshold_events = 1, change_state_on_threshold_event = 1, first_threshold_event_in = 5
```

Edit: this is not related to the problem, but the struct has some holes in it:
```
$ make TARGET=nrf52840 BOARD=dongle DEBUG=1
...
$ pahole -H 1 hello-world.nrf52840 | less
...
struct device_configuration {
        uint8_t                    type;                 /*     0     1 */
        char                       type_desc[10];        /*     1    10 */

        /* XXX 1 byte hole, try to pack */

        int16_t                    state_range_start;    /*    12     2 */
        int16_t                    state_range_end;      /*    14     2 */
        int16_t                    ini_value;            /*    16     2 */
        uint8_t                    send_threshold_events; /*    18     1 */
        uint8_t                    change_state_on_threshold_event; /*    19     1 */
        int16_t                    first_threshold_event_in; /*    20     2 */

        /* size: 22, cachelines: 1, members: 8 */
        /* sum members: 21, holes: 1, sum holes: 1 */
        /* last cacheline: 22 bytes */
};
...
```
Putting the biggest members first and gradually putting smaller members tends to avoid causing holes, and initializing the struct members by using designators (https://en.cppreference.com/w/c/language/struct_initialization) will allow reordering of struct fields without breaking code.

**Reply by rahav** [reply]
What you did above @pjonsson, opened a new line of thinking for me. Thank you !^10. It helped identify the problem, but I don't know how to solve it. I believe that this is general problem of **alignment of firmware in ROM** (variation in RAM do not affect the problem).

I ran a series of experiment identifying that ROM alignment is the culprit, as described below. If I'm correct, then my little saga is not important because this affects everyone.

# Experiment
I moved the simple code I provided, from my research project directory, to the `hello-world` directory. As was the case for you, it immediately started to work. Naturally I suspected a configuration in my research directory `project-conf.h` so I copied it to the `hello-world` directory. The modified `hello-world` stopped working.

I started to comment discrete parameters. After each change I checked the ROM and RAM requirements using the command: `arm-none-eabi-size build/nrf52840/dongle/hello-world.nrf52840`. Without exception, every parameter that affected the ROM caused the modified `hello-world` to work.

# To replicate the bug
To replicate, here is my `project-conf.h`. If you copy it to your directory, the dongle will stop working. If you comment, for example, `#define UIP_CONF_BUFFER_SIZE 200` it will run happily.

````
#ifndef PROJECT_CONF_H_
#define PROJECT_CONF_H_

#define UIP_CONF_BUFFER_SIZE 200

#ifndef NETSTACK_MAX_ROUTE_ENTRIES
#define NETSTACK_MAX_ROUTE_ENTRIES 32	// 32
#endif

#ifndef NBR_TABLE_CONF_MAX_NEIGHBORS
#define NBR_TABLE_CONF_MAX_NEIGHBORS 16	// 16
#endif

#define CSMA_CONF_MAX_NEIGHBOR_QUEUES 8	// 8

#define QUEUEBUF_CONF_NUM 16	// 16


/* ------------------------------------------------------------------------------------------------------------------ */
/*                                                  Multicast (MPL)                                                     */
/* ------------------------------------------------------------------------------------------------------------------ */
#ifndef UIP_MCAST6_CONF_ENGINE
#define UIP_MCAST6_CONF_ENGINE UIP_MCAST6_ENGINE_MPL
#endif

/* For Imin: Use 16 over CSMA, 64 over Contiki MAC */
#define ROLL_TM_CONF_IMIN_1         16
#define MPL_CONF_DATA_MESSAGE_IMIN  16
#define MPL_CONF_CONTROL_MESSAGE_IMIN  16

#define UIP_MCAST6_ROUTE_CONF_ROUTES 1

#define MPL_CONF_DOMAIN_SET_SIZE 5
#define MPL_CONF_SEED_SET_SIZE 10


#define DEPLOYMENT_MAPPING deployment_small7

#endif /* PROJECT_CONF_H_ */
````

**Reply by pjonsson** [reply]
Everything you mention sounds like some kind of memory corruption, but the risk that a modern GCC from ARM will be misaligning things seems really small. The nrf platform uses a more recent nrfx and various other components, does the code work on that platform?

If things are still broken on the nrf platform: my impression is that you're building in the hello-world directory and that does not use deployment, so start by removing everything that you can that still reproduces the problem.

**Reply by rahav** [reply]
WOW. I works well using the `nrf` platform. Thank you for replying today!

I forgot to include the `Makefile` that I used in my `hello-world` directory, below. Does it run for you, with above `project-conf.h` and below `Makefile` using the `nrf52840` platform?

````
CONTIKI_PROJECT = hello-world
all: $(CONTIKI_PROJECT)

PLATFORMS_EXCLUDE += sky z1

CONTIKI = ../..

include $(CONTIKI)/Makefile.identify-target
MODULES_REL += $(TARGET)
include $(CONTIKI)/Makefile.dir-variables
MODULES += $(CONTIKI_NG_NET_DIR)/ipv6/multicast
MODULES += os/services/shell   # RD added to enable Shell

include $(CONTIKI)/Makefile.include
````

**Reply by rahav** [reply]
Comes to think about it, once a program gets to a size that lead to memory corruption, we don't even need to play with Contiki's configuration file. The modified `hello-world.c`, along with the `Makefile`, and `project-conf.h` I provided above just happen to be once example. One of the source code in my research project, happens to be another.

Neither work when compiled with nrf52840 platform. To modify the ROM I simple added a `printf("\n");` and it ran with no issues.

**Reply by rahav** [reply]
I had the night to think about it. We don't really know that the different compiler and settings in the `nrf` platform is a solution, it just happens to work in this case. We also have examples, where adding printf("\n"); was not sufficient for the compiler to organized the code without memory corruption.

Do we have a compiler expert on the team, that can use above example to determine the core of the problem?

I am curisious to know whether it does not work on your system.

**Reply by pjonsson** [reply, marked as answer]
I don't know what people's backgrounds are, but I'm probably the compiler expert.

There's no question that you've discovered some kind of bug. The code snippet you started this thread with is rather short and simple, so it is probably not buggy. The data we have right now is that the code doesn't work on the nrf52840 platform, but it works on the nrf platform. As you point out, this could be coincidental, but it could also be because some bug is fixed in more recent versions of CMSIS, nrfx, or a slew of other things.

I'm not too keen on spending my spare time on the nrf52840 platform since the long term plan is that it will be superseded by the nrf platform. I haven't checked, but I think the SDK used for the nrf52840 platform is no longer supported by Nordic semi.

Since the nrf platform works for your example, is there something else that is preventing you from using that going forward?

**Reply by rahav** [reply]
Makes sense, I will move to `nrf`. Your advise is always sound. The difficulty will be that all scripts that run `make` will break. So far I did not find a way to specify the `NRF52_UPLOAD_SN` parameter with `nrf`, and without the serial number, J-Link will popup its UI.

**Reply by pjonsson** [reply]
The parameter is called `NRF_UPLOAD_SN` for the nrf platform.

**Reply by rahav** [reply]
THANK YOU!
